pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
 
 func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
 	if pl.enableReadWriteOncePod {
-		return pl.isReadWriteOncePodAccessModeConflict(pod)
+		return pl.isReadWriteOncePodAccessModeConflict(cycleState, pod)
 	}
+	p := &preFilterState{}
+	cycleState.Write(preFilterStateKey, p)
 	return framework.NewStatus(framework.Success)
 }
 
 // isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
 // This access mode restricts volume access to a single pod on a single node. Since only a single pod can
 // use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
-// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
-func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
 	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
 	if err != nil {
 		return framework.NewStatus(framework.Error, "error while getting node info")
 	}
+	pvcKeys, status := pl.keysForPVCWithRWOP(pod)
+	if status != nil {
+		return status
+	}
+	var nodesMatchPVCWithRWOP []*v1.Node
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeInfos[i]
+		pvcKeysNumber := uint32(len(pvcKeys))
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				atomic.AddUint32(&pvcKeysNumber, ^uint32(0))
```

Comment:
`pvcKeysNumber` is a local variable to the thread, you don't need atomic operations for it.

The problem is the `append` below

Reply:
I tried to fix it